### PR TITLE
[FlagGems Operator Development Competition] Add log10 pointwise operator

### PR DESCRIPTION
**Pull request by goldenfox2025**
## Summary

Add `log10` pointwise operator for the FlagGems Operator Development Competition.

This PR adds:

- `torch.log10`
- `torch.log10_`
- `torch.log10(..., out=...)`

## Implementation

- Added Triton pointwise implementation in `src/flag_gems/ops/log10.py`
- Registered `log10`, `log10_`, and `log10.out`
- Added correctness tests for forward / inplace / out variants
- Added benchmark coverage using the FlagGems built-in benchmark framework
- Added operator-specific benchmark shapes in `benchmark/core_shapes.yaml`

Formula used:

```python
log10(x) = log(x) * 0.4342944819032518
```

Implementation details:

- `pointwise_dynamic` + Triton JIT
- internal computation promoted to `float32`
- integer inputs promote to `float32`, aligned with PyTorch behavior
- validated on `float16`, `float32`, and `bfloat16`

## Accuracy Validation

Correctness tests cover:

- `log10`, `log10_`, `log10.out`
- integer input promotion behavior
- special values: `0`, `-0`, `-1`, `inf`, `-inf`, `nan`
- empty tensors: `(0,)`, `(4, 0)`, `(2, 0, 3)`
- non-contiguous input

All outputs are compared against PyTorch reference behavior.

## Benchmark

Command used:

```bash
PYTHONPATH=src python -m pytest benchmark/test_unary_pointwise_perf.py -k 'log10' -s --level comprehensive --record log
```

### Forward `log10`

| DType | Shape | PyTorch (ms) | FlagGems (ms) | Speedup |
| --- | --- | ---: | ---: | ---: |
| float16 | `(4096, 4096)` | 0.3656 | 0.3257 | 1.12x |
| float16 | `(64, 512, 512)` | 0.3226 | 0.3205 | 1.01x |
| float16 | `(1024, 65536)` | 1.3025 | 1.2872 | 1.01x |
| float32 | `(4096, 4096)` | 0.6467 | 0.6569 | 0.98x |
| float32 | `(64, 512, 512)` | 0.6400 | 0.6374 | 1.00x |
| float32 | `(1024, 65536)` | 2.5836 | 2.6153 | 0.99x |
| bfloat16 | `(4096, 4096)` | 0.3271 | 0.3256 | 1.00x |
| bfloat16 | `(64, 512, 512)` | 0.3251 | 0.3205 | 1.01x |
| bfloat16 | `(1024, 65536)` | 1.2954 | 1.2820 | 1.01x |

### Inplace `log10_`

| DType | Shape | PyTorch (ms) | FlagGems (ms) | Speedup |
| --- | --- | ---: | ---: | ---: |
| float16 | `(4096, 4096)` | 0.3195 | 0.3215 | 0.99x |
| float16 | `(64, 512, 512)` | 0.3164 | 0.3195 | 0.99x |
| float16 | `(1024, 65536)` | 1.2790 | 1.3158 | 0.97x |
| float32 | `(4096, 4096)` | 0.6400 | 0.6497 | 0.99x |
| float32 | `(64, 512, 512)` | 0.6502 | 0.6472 | 1.00x |
| float32 | `(1024, 65536)` | 2.7264 | 2.5805 | 1.06x |
| bfloat16 | `(4096, 4096)` | 0.3210 | 0.3195 | 1.00x |
| bfloat16 | `(64, 512, 512)` | 0.3174 | 0.3174 | 1.00x |
| bfloat16 | `(1024, 65536)` | 1.2718 | 1.3097 | 0.97x |

### Out variant `log10.out`

| DType | Shape | PyTorch (ms) | FlagGems (ms) | Speedup |
| --- | --- | ---: | ---: | ---: |
| float16 | `(4096, 4096)` | 0.3251 | 0.3215 | 1.01x |
| float16 | `(64, 512, 512)` | 0.3246 | 0.3236 | 1.00x |
| float16 | `(1024, 65536)` | 1.2841 | 1.3194 | 0.97x |
| float32 | `(4096, 4096)` | 0.6482 | 0.6339 | 1.02x |
| float32 | `(64, 512, 512)` | 0.6339 | 0.6574 | 0.96x |
| float32 | `(1024, 65536)` | 2.6537 | 2.7935 | 0.95x |
| bfloat16 | `(4096, 4096)` | 0.3226 | 0.3226 | 1.00x |
| bfloat16 | `(64, 512, 512)` | 0.3236 | 0.3215 | 1.01x |
| bfloat16 | `(1024, 65536)` | 1.3056 | 1.3179 | 0.99x |

## Test Coverage Checklist

- API branches: `log10`, `log10_`, `log10.out`
- Input scales: small / normal / large
- Input dims: covered through the unary pointwise shape set
- Special values: `0`, `-0`, negative, `nan`, `inf`, `-inf`
- Empty tensors: covered
- Tensor layout: contiguous and non-contiguous
- DTypes: `float16`, `float32`, `bfloat16`
- Integer behavior: promotion to `float32` verified against PyTorch

### CODE REVIEW

**Comment by you-and-you** [comment]
Resolve merge conflicts

**Comment by goldenfox2025** [comment]
> Resolve merge conflicts

Thanks for the reminder — the conflict has been resolved.

**Comment by you-and-you** [comment]
Please provide screenshots of the benchmark and correctness tests, rather than just tabular data.

**Comment by goldenfox2025** [comment]
image

image

image

image

image

**Comment by goldenfox2025** [comment]
> Please provide screenshots of the benchmark and correctness tests, rather than just tabular data.

It's been done.

**Comment by Miner-debug** [comment]
> tabular data

@you-and-you please review the screenshot

**Comment by you-and-you** [comment]
Resolve pre-commit

**Comment by goldenfox2025** [comment]
> Resolve pre-commit

Thanks for reminding me. It's been done.

**Review by huangyiqun — APPROVED** [review]
